LanguageFeatures/Patterns/type_inference_A33_t02.dart

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// If p with required type T is in an irrefutable context:
+/// - It is a compile-time error if M is not assignable to T.
+/// - Else if M is not a subtype of T then an implicit coercion or cast is
+///   inserted before the pattern binds the value, tests the value's type,
+///   destructures the value, or invokes a function with the value as a target
+///   or argument.
+///
+/// @description Check that if `M` is not a subtype of `T` then generic function
+/// instantiation is performed
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
+import "../../Utils/expect.dart";
+
+T foo<T>(T t) => t;
+
+main() {
+  var list = [foo];
+  var <T Function<T>(T)>[int Function(int) f1] = list;
+  Expect.equals(42, f1(42));
+  Expect.throws(() {
+    f1("1" as dynamic);
+  });
+
+  var <T Function<T>(T)>[int Function(int) fInt &&
+      String Function(String) fString] = list;
+  Expect.equals(42, fInt(42));
+  Expect.throws(() {
+    fInt("1" as dynamic);
+  });
+  Expect.equals("42", fString("42"));
+  Expect.throws(() {
+    fString(1 as dynamic);
+  });
+
+  final map = {"key1": foo};
+  final <String, int Function(int)>{"key1": f2} = map;
```

Comment:
Is it true that the existing version of line 45 just gives rise to a compile-time error because the initializing expression has a static type which is not assignable to the required type of the map pattern?

Reply:
Yes, this is true that the existing version of the line 54 produce a compile-time error. But updated version produce errors too. Is this an SDK issue?
```
unexpected analysis errors in type_inference_A33_t02.dart:
- Line 27, column 26: COMPILE_TIME_ERROR.PATTERN_TYPE_MISMATCH_IN_IRREFUTABLE_CONTEXT
  The matched value of type 'T Function<T>(T)' isn't assignable to the required type 'int Function(int)'.

- Line 33, column 26: COMPILE_TIME_ERROR.PATTERN_TYPE_MISMATCH_IN_IRREFUTABLE_CONTEXT
  The matched value of type 'T Function<T>(T)' isn't assignable to the required type 'int Function(int)'.

- Line 34, column 7: COMPILE_TIME_ERROR.PATTERN_TYPE_MISMATCH_IN_IRREFUTABLE_CONTEXT
  The matched value of type 'T Function<T>(T)' isn't assignable to the required type 'String Function(String)'.

- Line 45, column 44: COMPILE_TIME_ERROR.PATTERN_TYPE_MISMATCH_IN_IRREFUTABLE_CONTEXT
  The matched value of type 'T Function<T>(T)' isn't assignable to the required type 'int Function(int)'.
```

Reply:
Yes, the coercion should take place when line 45 has `<String, T Function<T>(T)>{"key1": int Function(int) f2}`. However, line 52 should be able to work in the previous version, using the type `int Function(int)` for `x` (otherwise we won't have the coercion at all, and it should work to do it like that).

In both cases, it looks like the CFE just does not insert coercions, so it makes no difference whether the type difference occurs in the correct place or a different place. ;-)

However, I'm arguing in https://github.com/dart-lang/language/issues/2857#issuecomment-1458854384 that we should drop these "hidden coercions" entirely because they are a rather bad fit for various generalizations, in particular for one generalization which has been mentioned as desirable by members of the language team several times: Formal parameters expressed via patterns.

So we'll see how it goes, but for now we should stick to the current specification, which means that we should get the coercions as mentioned above.